ngraph/frontend/paddlepaddle/src/pdpd_utils.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "pdpd_utils.hpp"
+
+namespace ngraph
+{
+    namespace frontend
+    {
+        namespace pdpd
+        {
+            std::shared_ptr<TensorPlacePDPD> castToTensorPlace(const Place::Ptr& place)
+            {
+                if (auto var_place = std::dynamic_pointer_cast<TensorPlacePDPD>(place))
+                {
+                    return var_place;
+                }
+                else if (auto in_port_place = std::dynamic_pointer_cast<InPortPlacePDPD>(place))
+                {
+                    return in_port_place->get_source_tensor_pdpd();
+                }
+                else if (auto out_port_place = std::dynamic_pointer_cast<OutPortPlacePDPD>(place))
+                {
+                    return out_port_place->get_target_tensor_pdpd();
+                }
+                FRONT_END_GENERAL_CHECK(false, "Cannot cast this Place to TensorPlacePDPD.");
+            }
+
+            void traverse_down(const std::vector<Place::Ptr>& start_nodes,
+                               std::vector<std::shared_ptr<OpPlacePDPD>>* ordered_ops,
+                               std::vector<std::shared_ptr<TensorPlacePDPD>>* ordered_tensors)
+            {
+                std::queue<OpPlacePDPD*> q;
+                std::unordered_set<OpPlacePDPD*> visited;
+
+                auto check_and_update = [&](const std::shared_ptr<OpPlacePDPD>& op) -> bool {
+                    if (op && !visited.count(op.get()))
+                    {
+                        visited.insert(op.get());
+                        q.push(op.get());
+                        if (ordered_ops)
+                            ordered_ops->push_back(op);
+                        return true;
+                    }
+                    return false;
+                };
+
+                for (const auto& node : start_nodes)
+                {
```

Comment:
I believe this cycle should just 'push to queue'. Attributes like 'visited' and ordered_ops will be filled in next cycle
Looks like there is some duplication here.
And also algorithm doesn't look like a topological sorting

Reply:
"also algorithm doesn't look like a topological sorting" - `traverse_up` is a topological sorting (tensor place has only one parent, input place is also has only one parent) 
`traverse_down` is not a topological sorting, but I believe it's unnecessary for our purpose

"I believe this cycle should just 'push to queue'. Attributes like 'visited' and ordered_ops will be filled in next cycle" - 
I guess we should do the same checks as in main loop